### PR TITLE
feat(49962): Disallow comparison against NaN 

### LINKED ISSUE #49962: Disallow comparison against `NaN`
# Suggestion

## 🔍 Search Terms

disallow NaN
comparing with NaN
disallow comparing NaN

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [?] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* ✅ This wouldn't change the runtime behavior of existing JavaScript code
* ✅ This could be implemented without emitting different JS based on the types of the expressions
* ✅ This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* ✅ This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Similar to https://github.com/microsoft/TypeScript/pull/45978 . So the error should be something like `The condition will always return 'false' since NaN !== NaN. Use Number.isNaN.`

## 📃 Motivating Example

```typescript
function setReputation (n: number)
{
  // ! here is the bug:
  if (n === NaN) throw new Error('setReputation:: Impossible reputation')
  // update the record in a remote DB
}
const foundGolems = 0
const killedGolems = 0
setReputation(killedGolems / foundGolems * 100)
// 0/0 is NaN
```
or what has actually happened to me:
```typescript
export function clamp ( // @license CC0
	n:   number,
	min: number|bigint,
	max: number|bigint,
): typeof n
export function clamp (
	n:   bigint,
	min: number|bigint,
	max: number|bigint,
): typeof n
export function clamp (
	n:   number|bigint,
	min: number|bigint,
	max: number|bigint,
): typeof n
{
	// +null === 0
	// +'' === 0
	// +undefined === NaN
	const nFixed = typeof n === 'bigint' ? n : (
		// @ts-ignore
		'' === n || n === null || typeof n === 'symbol'
		? NaN : +n
	)
	const minFixed = typeof min === 'bigint' ? min : (
		// @ts-ignore
		'' === min || min === null || typeof min === 'symbol'
		? NaN : +min
	)
	const maxFixed = typeof max === 'bigint' ? max : (
		// @ts-ignore
		'' === max || max === null || typeof max === 'symbol'
		? NaN : +max
	)
	if (nFixed === NaN)
	{
		const ret = nFixed
		//const ret = min == max ? (typeof min === 'bigint' ? +min.toString() : min) : nFixed
		console.warn(
			'clamp(n=%O, min=%O, max=%O):: n === NaN; returning %O',
			n, min, max, ret,
		)
		return ret
	}
	let ret: typeof n
	if (
		minFixed === NaN
		&&
		maxFixed === NaN
	)
	{
		ret = nFixed
		console.warn(
			'clamp(n=%O, min=%O, max=%O):: min === NaN && max === NaN; returning `n` (%O)',
			n, min, max, ret,
		)
	}
	else if (maxFixed < minFixed)
	{
		// throw new RangeError
		//ret = maxFixed // similar to torch.clamp
		ret = minFixed // because arguments are specified in the left-to-right *order*.
		//ret = nFixed < maxFixed ? maxFixed : (nFixed > minFixed ? minFixed : nFixed)
		/*console.warn(
			'clamp(n=%O, min=%O, max=%O):: max<min; returning `min` (%O)',
			n, min, max, ret
		)*/
	}
	else
	{
		ret = nFixed < minFixed ? minFixed : (nFixed > maxFixed ? maxFixed : nFixed)
	}
	if (typeof n === 'bigint')
	{
		return typeof ret === 'bigint' ? ret : BigInt(Math.floor(ret))
	}
	if (typeof ret === 'bigint')
	{
		return +ret.toString()
	}
	return ret
}
```

## 💻 Use Cases

Since NaN is very rare, a programmer could forget to use `Number.isNaN(variable)`, and accidentally use `===`.
Or a programmer could have never even learned this, and just applies knowledge from other langs where NaN === NaN.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #49962

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -34618,6 +34619,23 @@ namespace ts {
...
+            function checkNaNEquality(errorNode: Node | undefined, operator: SyntaxKind, left: Expression, right: Expression, leftType: Type, rightType: Type) {
+                const isLeftNaN = leftType.flags & TypeFlags.Number && isNaN(skipParentheses(left));
+                const isRightNaN = rightType.flags & TypeFlags.Number && isNaN(skipParentheses(right));
+                if (isLeftNaN || isRightNaN) {
+                    const err = error(errorNode, Diagnostics.This_condition_will_always_return_0,
+                        tokenToString(operator === SyntaxKind.EqualsEqualsEqualsToken || operator === SyntaxKind.EqualsEqualsToken ? SyntaxKind.FalseKeyword : SyntaxKind.TrueKeyword));
+                    if (isLeftNaN && isRightNaN) return;
+                    const location = isLeftNaN ? right : left;
+                    addRelatedInfo(err, createDiagnosticForNode(location, Diagnostics.Did_you_mean_0,
+                        `${operator === SyntaxKind.ExclamationEqualsEqualsToken || operator === SyntaxKind.ExclamationEqualsToken ? tokenToString(SyntaxKind.ExclamationToken) : ""}Number.isNaN(${getTextOfNode(location)})`));
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:34632]: If the expression is too complicated, you might want to swap to using `...`.

So if you can, use

```ts
const location = skipParentheses(isLeftNaN ? right : left);
const locationString = isEntityNameExpression(location) ? entityNameToString(location) : "...";
addRelatedInfo(...);
```

Comment (a-tarasyuk) [reply src/compiler/checker.ts:34632]: What do you think about using the `defaultMaximumTruncationLength`? I think suggestions like Number.isNaN(a[0]) would be better than `Number.isNaN(....)`. Is it better to only allow this suggestion for `EntityNameExpression`s?

```ts
const expression = isLeftNaN ? right : left;
const expressionText = getTextOfNode(skipParentheses(expression), /*includeTrivia*/ false);
addRelatedInfo(err, createDiagnosticForNode(expression, Diagnostics.Did_you_mean_0,
    `${operator === SyntaxKind.ExclamationEqualsEqualsToken || operator === SyntaxKind.ExclamationEqualsToken
        ? tokenToString(SyntaxKind.ExclamationToken)
        : ""}Number.isNaN(${expressionText.length > defaultMaximumTruncationLength ? "..." : expressionText})`));
```

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:34632]: I think it's fine to just limit it to `EntityNameExpression`s. The problem is not necessarily just about length, it's also about multi-line code and indentation too.

**Comment by DanielRosenwasser** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot perf test this faster

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -34618,6 +34619,25 @@ namespace ts {
...
+                    const operatorString = operator === SyntaxKind.ExclamationEqualsEqualsToken || operator === SyntaxKind.ExclamationEqualsToken ? tokenToString(SyntaxKind.ExclamationToken) : "";
+                    const location = isLeftNaN ? right : left;
+                    const expression = skipParentheses(location);
+                    addRelatedInfo(err, createDiagnosticForNode(location, Diagnostics.Did_you_mean_0,
+                        `${operatorString}Number.isNaN(${isEntityNameExpression(expression) ? entityNameToString(expression) : "..."})`));
+                }
+            }
+
+            function isNaN(expr: Expression): boolean {
+                return isIdentifier(expr) && expr.escapedText === "NaN";
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:34639]: Currently this check is too lax - what you'll need to do is try to resolve the global `NaN` value up-front. If it's present, then you'll need to check whether this identifier resolves to that symbol.

You should have a test like

```ts
function weirdIsNaN(value: number, NaN: number) {
    return value === NaN;
}

function weirdNaNIsNaN(NaN: number) {
    return NaN === NaN;
}
```

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
Weird that the perf test failed?

@typescript-bot perf test this faster

**Comment by DanielRosenwasser** [comment]
Experimenting with the perf test issues over at https://github.com/microsoft/TypeScript/pull/50812

**Comment by DanielRosenwasser** [comment]
@navya9singh @weswigham can you try to take a quick look? I think this is in a pretty good state, but I'd like to get at least one more pair of eyes by end-of-day before merging.

**Comment by weswigham** [comment]
@typescript-bot perf test this faster

**Review by weswigham — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
I think the perf test looks fine. I think we're good to merge!

**Pull request merged**